### PR TITLE
Convert Config, Camera, Hand, Water, Sky and Info to locator services

### PR DESCRIPTION
**Pull request by bwrsandman**
Supersedes
* #778 
* #779 
* #780
* #783 

This refactor speeds up parsing compilation by 2% on ubuntu clang debug, on mac debug, the parsing time drops by 30%. It also removes Game.h as one of the top time spent including header.

### CODE REVIEW

File: src/Debug/Gui.cpp [hunk 1]
```diff
@@ -658,7 +662,7 @@ void Gui::RenderArrow(const std::string& name, const ImVec2& pos, const ImVec2&
 	ImGui::SetNextWindowSize(size);
 	if (ImGui::Begin((name + " Frame").c_str(), nullptr, boxOverlayFlags))
 	{
-		std::string strId = name + " Arrow";
+		const std::string strId = name + " Arrow";
```
Comment (Daniels118) [inline src/Debug/Gui.cpp:665]:
```suggestion
		const auto strId = name + " Arrow";
```

File: src/Debug/Console.cpp [hunk 2]
```diff
@@ -215,9 +217,9 @@ int Console::InputTextCallback(ImGuiInputTextCallbackData* data)
 	return 0;
 }
 
-void Console::Draw(Game& game)
+void Console::Draw() noexcept
 {
-	ImGuiIO& io = ImGui::GetIO();
+	const ImGuiIO& io = ImGui::GetIO();
```
Comment (Daniels118) [inline src/Debug/Console.cpp:222]:
```suggestion
	const auto& io = ImGui::GetIO();
```

File: src/Debug/Console.cpp [hunk 3]
```diff
@@ -273,14 +275,14 @@ void Console::Draw(Game& game)
 		_items.clear();
 	}
 	ImGui::SameLine();
-	bool copyToClipboard = ImGui::SmallButton("Copy");
+	const bool copyToClipboard = ImGui::SmallButton("Copy");
```
Comment (Daniels118) [inline src/Debug/Console.cpp:278]:
```suggestion
	const auto copyToClipboard = ImGui::SmallButton("Copy");
```

File: src/Debug/Console.cpp [hunk 4]
```diff
@@ -273,14 +275,14 @@ void Console::Draw(Game& game)
...
 
 	ImGui::Separator();
 
-	const float footerHeightToReserve =
-	    ImGui::GetStyle().ItemSpacing.y + ImGui::GetFrameHeightWithSpacing(); // 1 separator, 1 input text
-	ImGui::BeginChild("ScrollingRegion", ImVec2(0, -footerHeightToReserve), false,
-	                  ImGuiWindowFlags_HorizontalScrollbar); // Leave room for 1 separator + 1 InputText
+	// 1 separator, 1 input text
+	const float footerHeightToReserve = ImGui::GetStyle().ItemSpacing.y + ImGui::GetFrameHeightWithSpacing();
```
Comment (Daniels118) [inline src/Debug/Console.cpp:283]:
```suggestion
	const auto footerHeightToReserve = ImGui::GetStyle().ItemSpacing.y + ImGui::GetFrameHeightWithSpacing();
```

File: src/Debug/Gui.cpp [hunk 5]
```diff
@@ -200,10 +201,12 @@ bool Gui::ProcessEvents(const SDL_Event& event) noexcept
 
 	ImGui_ImplSDL2_ProcessEvent(&event);
 
-	ImGuiIO& io = ImGui::GetIO();
+	const ImGuiIO& io = ImGui::GetIO();
```
Comment (Daniels118) [inline src/Debug/Gui.cpp:204]:
```suggestion
	const auto& io = ImGui::GetIO();
```

File: src/Debug/Gui.cpp [hunk 6]
```diff
@@ -224,7 +227,7 @@ bool Gui::ProcessEvents(const SDL_Event& event) noexcept
 	return _stealsFocus;
 }
 
-bool Gui::CreateFontsTextureBgfx()
+bool Gui::CreateFontsTextureBgfx() noexcept
 {
 	// Build texture atlas
 	ImGuiIO& io = ImGui::GetIO();
```
Comment (Daniels118) [inline src/Debug/Gui.cpp:233]: Can be declared as `const auto`?

File: src/Debug/Gui.cpp [hunk 7]
```diff
@@ -245,10 +248,10 @@ bool Gui::CreateFontsTextureBgfx()
 	return true;
 }
 
-bool Gui::CreateDeviceObjectsBgfx()
+bool Gui::CreateDeviceObjectsBgfx() noexcept
 {
 	// Create shaders
-	bgfx::RendererType::Enum type = bgfx::getRendererType();
+	const bgfx::RendererType::Enum type = bgfx::getRendererType();
```
Comment (Daniels118) [inline src/Debug/Gui.cpp:254]: Can use `auto` type?

File: src/Debug/Gui.cpp [hunk 8]
```diff
@@ -565,7 +569,7 @@ bool Gui::ShowMenu(Game& game)
 
 			if (ImGui::BeginMenu("Field of View"))
 			{
-				auto& camera = game.GetCamera();
+				auto& camera = Locator::camera::value();
 				float fieldOfView = glm::degrees(camera.GetHorizontalFieldOfView());
```
Comment (Daniels118) [inline src/Debug/Gui.cpp:573]: Can be declared as `const auto`?

**Review by Daniels118 — COMMENTED** [review]
Using `auto` where applicable.

**Review by Daniels118 — CHANGES_REQUESTED** [review]
What's the purpose of 23d7dfe235bb7b113fb64d3e5cd2648ab63ab897? It improves the code readability a bit, but it could increase the load time on low-end devices. If you want to keep the improved readability, maybe it would be better to change the type of `_infoConstants` to `unique_ptr`.

**Comment by bwrsandman** [comment]
It's prep for the following commit which makes it into a `shared_ptr` through the locator service. The 0.5MB copy issue is also removed in the next commit.

**Review by Daniels118 — APPROVED** [review]

**Review by raffclar — APPROVED** [review]

**Comment by raffclar** [comment]
Really nice work. One step closer to `Game::Instance()` being removed

**Pull request merged**